File: app/src/main/java/com/veganbeauty/app/features/community/notification/CommunityNotificationHelper.java
package com.veganbeauty.app.features.community.notification;

import android.content.Context;

import com.veganbeauty.app.data.repository.CommunityNotificationRepository;
import com.veganbeauty.app.utils.RootieBrandHelper;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class CommunityNotificationHelper {

    public static void addCommunityNotificationForUser(
            Context context,
            String targetUserId,
            String id,
            String userId,
            String userName,
            String userAvatar,
            String type,
            String actionType,
            String content,
            String postId,
            String commentId
    ) {
        if (CommunityNotificationRepository.isGuest(context)) return;

        if (targetUserId == null || targetUserId.isEmpty()) return;

        String currentDate = new SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).format(new Date());
        String currentTime = new SimpleDateFormat("HH:mm", Locale.getDefault()).format(new Date());

        java.util.Map<String, Object> notificationMap = new java.util.HashMap<>();
        notificationMap.put("id", id);
        notificationMap.put("userId", userId != null ? userId : "");
        notificationMap.put("userName", userName);
        notificationMap.put("userAvatar", userAvatar != null ? userAvatar : RootieBrandHelper.AVATAR_URL);
        notificationMap.put("type", type);
        notificationMap.put("actionType", actionType);
        notificationMap.put("content", content);
        notificationMap.put("time", currentTime);
        notificationMap.put("date", currentDate);
        notificationMap.put("isRead", false);
        notificationMap.put("postId", postId != null ? postId : "");
        notificationMap.put("commentId", commentId != null ? commentId : "");
        
        // Asynchronously add to firestore so it doesn't block UI thread
        new Thread(() -> {
            new com.veganbeauty.app.data.remote.FirestoreService().addCommunityNotification(targetUserId, notificationMap);
        }).start();
    }

    public static void addCommunityNotificationLocalOnly(
            Context context,
            String targetUserId,
            String id,
            String userId,
            String userName,
            String userAvatar,
            String type,
            String actionType,
            String content,
            String postId,
            String commentId
    ) {
        if (CommunityNotificationRepository.isGuest(context)) return;
        if (targetUserId == null || targetUserId.isEmpty()) return;

        File file = new File(context.getFilesDir(), "local_notifications_" + targetUserId + ".json");
        List<ComNotificationItem> list = new ArrayList<>();

        if (file.exists()) {
            try (java.io.BufferedReader reader = new java.io.BufferedReader(new java.io.InputStreamReader(new java.io.FileInputStream(file), java.nio.charset.StandardCharsets.UTF_8))) {
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    sb.append(line);
                }
                org.json.JSONArray jsonArray = new org.json.JSONArray(sb.toString());
                for (int i = 0; i < jsonArray.length(); i++) {
                    org.json.JSONObject obj = jsonArray.getJSONObject(i);
                    list.add(new ComNotificationItem(
                            obj.getString("id"),
                            obj.optString("userId"),
                            obj.getString("userName"),
                            obj.optString("userAvatar"),
                            obj.getString("type"),
                            obj.getString("actionType"),
                            obj.getString("content"),
                            obj.getString("time"),
                            obj.getString("date"),
                            obj.optBoolean("isRead", false),
                            obj.optString("postId"),
                            obj.optString("commentId"),
                            ""
                    ));
                }
            } catch (Exception e) {}
        }

        String currentDate = new java.text.SimpleDateFormat("dd/MM/yyyy", java.util.Locale.getDefault()).format(new java.util.Date());
        String currentTime = new java.text.SimpleDateFormat("HH:mm", java.util.Locale.getDefault()).format(new java.util.Date());

        ComNotificationItem newItem = new ComNotificationItem(
                id, userId, userName, userAvatar, type, actionType, content,
                currentTime, currentDate, false, postId, commentId, ""
        );

        list.removeIf(it -> it.getId().equals(id));
        list.add(0, newItem);

        try {
            org.json.JSONArray newArray = new org.json.JSONArray();
            for (ComNotificationItem item : list) {
                org.json.JSONObject obj = new org.json.JSONObject();
                obj.put("id", item.getId());
                obj.put("userId", item.getUserId() != null ? item.getUserId() : "");
                obj.put("userName", item.getUserName());
                obj.put("userAvatar", item.getUserAvatar() != null ? item.getUserAvatar() : "");
                obj.put("type", item.getType());
                obj.put("actionType", item.getActionType());
                obj.put("content", item.getContent());
                obj.put("time", item.getTime());
                obj.put("date", item.getDate());
                obj.put("isRead", item.isRead());
                obj.put("postId", item.getPostId() != null ? item.getPostId() : "");
                obj.put("commentId", item.getCommentId() != null ? item.getCommentId() : "");
                newArray.put(obj);
            }
            try (java.io.FileOutputStream fos = new java.io.FileOutputStream(file)) {
                fos.write(newArray.toString().getBytes(java.nio.charset.StandardCharsets.UTF_8));
            }
            CommunityNotificationRepository.getInstance(context).refresh();
        } catch (Exception e) {}
    }
}

File: app/src/main/java/com/veganbeauty/app/features/community/notification/CommunityNotificationHelper.kt
package com.veganbeauty.app.features.community.notification

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object CommunityNotificationHelper {

    @JvmStatic
    fun addCommunityNotification(
        context: Context,
        id: String,
        userId: String?,
        userName: String,
        userAvatar: String?,
        type: String, // "POST", "INTERACTION", "ORDER"
        actionType: String,
        content: String,
        postId: String? = null,
        commentId: String? = null
    ) {
        val file = File(context.filesDir, "local_notifications.json")
        val list = mutableListOf<ComNotificationItem>()
        
        // Load existing notifications
        if (file.exists()) {
            try {
                val jsonArray = JSONArray(file.readText())
                for (i in 0 until jsonArray.length()) {
                    val obj = jsonArray.getJSONObject(i)
                    list.add(
                        ComNotificationItem(
                            id = obj.getString("id"),
                            userId = obj.optString("userId").takeIf { it.isNotEmpty() && it != "null" },
                            userName = obj.getString("userName"),
                            userAvatar = obj.optString("userAvatar").takeIf { it.isNotEmpty() && it != "null" },
                            type = obj.getString("type"),
                            actionType = obj.getString("actionType"),
                            content = obj.getString("content"),
                            time = obj.getString("time"),
                            date = obj.getString("date"),
                            isRead = obj.optBoolean("isRead", false),
                            postId = obj.optString("postId").takeIf { it.isNotEmpty() && it != "null" },
                            commentId = obj.optString("commentId").takeIf { it.isNotEmpty() && it != "null" }
                        )
                    )
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        } else {
            // Load initial from assets if local doesn't exist yet
            try {
                val jsonString = context.assets.open("notification_com.json").bufferedReader().use { it.readText() }
                val jsonArray = JSONArray(jsonString)
                for (i in 0 until jsonArray.length()) {
                    val obj = jsonArray.getJSONObject(i)
                    list.add(
                        ComNotificationItem(
                            id = obj.getString("id"),
                            userId = obj.optString("userId").takeIf { it.isNotEmpty() && it != "null" },
                            userName = obj.optString("userName", "Hệ thống"),
                            userAvatar = obj.optString("userAvatar", ""),
                            type = obj.getString("type"),
                            actionType = obj.getString("actionType"),
                            content = obj.getString("content"),
                            time = obj.getString("time"),
                            date = obj.getString("date"),
                            isRead = obj.optBoolean("isRead", false),
                            postId = obj.optString("postId").takeIf { it.isNotEmpty() && it != "null" },
                            commentId = obj.optString("commentId").takeIf { it.isNotEmpty() && it != "null" }
                        )
                    )
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        // Create new item
        val currentDate = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).format(Date())
        val currentTime = SimpleDateFormat("HH:mm", Locale.getDefault()).format(Date())

        val newItem = ComNotificationItem(
            id = id,
            userId = userId,
            userName = userName,
            userAvatar = userAvatar ?: "https://res.cloudinary.com/dpjkzxjl2/image/upload/v1780560866/Rootie_logo.png",
            type = type,
            actionType = actionType,
            content = content,
            time = currentTime,
            date = currentDate,
            isRead = false,
            postId = postId,
            commentId = commentId
        )

        // Prevent duplicate
        list.removeAll { it.id == id }
        list.add(0, newItem)

        // Save back
        try {
            val newArray = JSONArray()
            for (item in list) {
                val obj = JSONObject().apply {
                    put("id", item.id)
                    put("userId", item.userId ?: "")
                    put("userName", item.userName)
                    put("userAvatar", item.userAvatar ?: "")
                    put("type", item.type)
                    put("actionType", item.actionType)
                    put("content", item.content)
                    put("time", item.time)
                    put("date", item.date)
                    put("isRead", item.isRead)
                    put("postId", item.postId ?: "")
                    put("commentId", item.commentId ?: "")
                }
                newArray.put(obj)
            }
            file.writeText(newArray.toString())
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
